### DISCUSSION TITLE
NeuroDock Issues

### DISCUSSION THREAD

**Discussion by mmsarhan** [muscbridge/PyDKE#280] [Technical Support] [answered]
I am having trouble Running NeuroDock on my computer witch is a Windows 10 64-bit Thinkpad. I was able to download the Docker application, but keep running into a problem when I try to download NeuroDock. This is the error message I get:
Error response from daemon: manifest for dmri/neurodock:v0.2 not found: manifest unknown: manifest unknown

**Comment by TheJaeger** [comment]
`dmri/neurodock:v0.2` is extremely old and it is most likely not the version you pulled. Try running `dmri/neurodock:latest` instead.

**Reply by mmsarhan** [reply]
@TheJaeger  
I realized that the tag name was latest and not v0.2 anymore and was able to download everything, but I was not able to figure out how too use it to process data. The instructions were very confusing and I could not figure it out. I am using DICOM folder with all the DKI images, and the instructions use a nifti folder. Is there a different way to process the data with a dicom folder?

**Reply by TheJaeger** [reply]
PyDesigner works with DICOM folders too as long as they're sorted into sequences (one sequence per folder). In that case, you may provide the path to DICOM folder directly.

Otherwise, you can use Chris Rorden's [dcm2niix](https://github.com/rordenlab/dcm2niix) to convert your DICOM files to NifTi files. I recommend this route.

**Reply by mmsarhan** [reply]
If I wanted to stay with the dicom folder is this the correct way to set up the command to process the data?

docker run -it --rm -v /Users/Tash/OneDrive/Documents/(DKI folder):/data \
    dmri/neurodock pydesigner --standard \
    --output /data/processed \
    /data/nifti/bond_dwi.nii,/data/nifti/bond_topup.nii(What do I put instead of this line)

**Reply by TheJaeger** [reply]
You would simply replace it with the path to folder containing DKI dicoms. But I also recommend changing your mount path to `/Users/Tash/OneDrive/Documents/`. Your full command would then be:

```
docker run -it --rm -v /Users/Tash/OneDrive/Documents:/data dmri/neurodock:latest pydesigner --standard --output /data/processed /data/(DKI folder)
```

**Reply by mmsarhan** [reply]
why do you think I should change the month path?

**Reply by TheJaeger** [reply]
It's so you could input the path to your DKI folder relative to mount point `/data` much more easily. For example, if you mount your DICOM folder directly to `/data` with `/Users/Tash/OneDrive/Documents/(DKI folder):/data `, then `/data` becomes DKI folder and that's the only directory Docker will have access to. This would be a problem if your output directly is nestled outside of the folder containing your DICOMs because it wouldn't be accessible anymore. Hence why I recommend mounting one directory above your target directory.

**Reply by mmsarhan** [reply]
This is what I got when I put in the command. Did I do something wrong?
![Screenshot (4)](https://user-images.githubusercontent.com/95373490/148416574-fc32ed64-4105-4c78-a339-8f40e7b31572.png)

**Reply by TheJaeger** [reply]
Yeah, the container is specified incorrectly. You need to parse in `dmri/neurodock:latest`.

**Reply by TheJaeger** [reply]
The other issue you'll come across is the way your directory is named. There's a period in your directory, which will break pathing. You need to replace `2.5mm` with something valid like `2_5mm`

**Reply by mmsarhan** [reply]
By Parsing do you mean that it needs to broken into two separate commands?

**Reply by mmsarhan** [reply]
It keeps saying the path I am using is not a windows path.
![Screenshot (5)](https://user-images.githubusercontent.com/95373490/148422131-a29b8c72-09e2-497b-9c4a-5636040caa43.png)

**Reply by TheJaeger** [reply]
Try mounting with a full path that includes the drive i.e. `C:\Users\Tash\OneDrive\Documents`

**Comment by mmsarhan** [comment, marked as answer]
It is very close to running but keeps running into an issue. It keeps saying that it find the neurodock image but I have the Docker Desktop running.
![Screenshot (6)](https://user-images.githubusercontent.com/95373490/148808740-fdcbde41-2534-43cf-a064-cf8071954876.png)

**Reply by mmsarhan** [reply]
I realized I misspelled dmri in the command, but now it will not run because of something in the pydesigner container. This is what I get when I run the command
![Screenshot (7)](https://user-images.githubusercontent.com/95373490/148810786-0a42b6e9-6c43-4422-a258-75528c3179e5.png)
.

**Reply by TheJaeger** [reply]
PyDesigner cannot find the input path specified. You have to be careful when specifying your Docker mount - make sure that the path is valid and exists. Converting your mount's path `/data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20` to system path would yield `C:\Users\Tash\Documents\DKI_BIPOLAR_2_5_mm_64dir_50slices_20`. Is this where your input images are stored on the system?

**Reply by mmsarhan** [reply]
I made sure both the mounted path and the input path is correct but it still yeilded the same result and I am unsure why. Yes the path to the folder is C:\Users\Tash\Onedrive\Documents\DKI_BIPOLAR_2_5mm_64dir_50slices_20
![Screenshot (8)](https://user-images.githubusercontent.com/95373490/148998036-c2dea912-f34e-471e-af81-56dd387f290d.png)

**Reply by TheJaeger** [reply]
Try specifying your Windows path in Docker mount with double slashes i.e. `C:\\Users\\Tash\\Onedrive\\Documents`

**Reply by mmsarhan** [reply]
That does not work either. I know the path is correct, but for some reason it will not work and I am unsure why that is. Is it something with the output?

**Reply by TheJaeger** [reply]
Run this command and paste the output here

```
docker run -it --rm -v C:\\Users\\Tash\\Onedrive\\Documents:/data dmri/neurodock:latest ls /data
```

**Reply by mmsarhan** [reply]
is this the correct command to run on a windows computer.
docker run -it --rm -v C:\\Users\\Tash\\Onedrive\\Documents:/data dmri/neurodock:latest ls /data C:\\Users\\Tash\\Onedrive\\Documents\\DKI_BIPOLAR_2_5_mm_64dir_50slices_20

**Reply by TheJaeger** [reply]
No, you'd run the command as-is. It lists what the Docker container sees in `/data`. We're doing this to ensure the path is properly mounted.

**Reply by mmsarhan** [reply]
This is the output it gave me when I ran the command.

C:\Users\Tash>docker run -it --rm -v C:\\Users\\Tash\\Onedrive\\Documents:/data dmri/neurodock:latest ls /data
 111_CR1_V1_EVS.csv        'COCA Receipts'
 111_CR1_V2_Post_RESP.csv   COCA_Scanned_Subjects.xlsx
 111_CR2_V2_Post_RESP.csv  'Custom Office Templates'
 111_CR2_V2_Pre_EVS.csv     DKI_BIPOLAR_2_5mm_64dir_50slices_20
 113_CR1_V2_Post_RESP.csv   HealthHistory_COCA.docx
 113_CR2_V2_Pre_RESP.csv    IBM
 114_CR2_V1_EVS.csv        'Missed Participants.xlsx'
 115_CR1_V1_RESP.csv        MobaXterm
 115_CR1_V2_Post_EVS.csv   'My Data Sources'
 118_CR1_V2_Pre_RESP.csv   'My Experiment'
 120_CR1_V1_RESP.csv       'My Experiments'
 123_CR1_V2_Pre_RESP.csv   'My Maps'
 123_CR2_V1_EVS.csv        'OnsetFile_Chart (Autosaved).xlsm'
 123_CR2_V2_Post_RESP.csv   OnsetFile_Chart.xlsx
 124_CR1_V1_EVS.csv         OnsetFile_Chart_114_CR1_PreTMS.xlsm
 124_CR1_V2_Pre_RESP.csv    OnsetFile_Chart_115_CR1_V1.xlsx
 124_CR2_V1_EVS.csv        'Python Scripts'
 126_CR1_V1_RESP.csv        R
 130_CR1_V1.csv             Receipts
 130_CR2_V1_RESP.csv       'Response_Chart(4).xlsx'
 133_CR2_V1_EVS.csv         Response_Chart.xlsx
 134_CR2_V1_EVS.csv         SigmaPlot
 136_CR2_V2_Post_EVS.csv    Zoom
 137_CR1_V2_Pre_EVS.csv     desktop.ini
'CDAP info.docx'           '~$LT_tDCS.docx'
'COCA Docs'                '~$scussion_Full_1.docx'
'COCA E-Prime'             '~$test_LL_sess.xlsx'

C:\Users\Tash>

**Reply by mmsarhan** [reply]
![Screenshot (9)](https://user-images.githubusercontent.com/95373490/149177202-2c11b250-09d8-4a51-a9dd-76f849624f6a.png)

**Reply by TheJaeger** [reply]
The folder ` DKI_BIPOLAR_2_5mm_64dir_50slices_20` definitely exists and visible to NeuroDock. The problem is the way your input path is specified. The container runs on an Ubuntu-based so paths relative to `/data` have to be specified in Linux fashion i.e. forward slashes instead of backslashes.

The full command to process this specific dataset will be

```
docker run -it --rm -v C:\Users\Tash\Onedrive\Documents:/data dmri/neurodock:latest pydesigner --standard --output /data/output /data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20
```

In your previous command, you specified the path to your input image as `C:\Users\Tash\Onedrive\Documents\DKI_BIPOLAR_2_5mm_64dir_50slices_20`, which does not exist for the container. Because you mounted `C:\Users\Tash\Onedrive\Documents` to `/data`, all paths have to be relative to `/data` and specified with forward slashes.

You'd also need to create a folder at the path `C:\Users\Tash\Onedrive\Documents\output` since you specified an output folder. The output folder needs to exist to save all relevant processing files.

**Reply by mmsarhan** [reply]
So the mounted file is in windows but everything after that is relative to the /data container. I ran the command just like you gave it to me and it still did not work. Not sure why this keeps happening.
![Screenshot (10)](https://user-images.githubusercontent.com/95373490/149191482-4e8d4778-7c34-49c4-b455-e9b925f1fce7.png)
.

**Reply by TheJaeger** [reply]
This is definitely weird. I tried replicating the issue using the same container version on a DICOM directory (Windows as well) and it works just fine on my end. The output should look like the image attached.

![Screenshot 2022-01-12 123354](https://user-images.githubusercontent.com/13654344/149192377-17cd94ef-20ec-4c4b-bc25-9e1fc0b0f3a2.jpg)

One more thing we can try is running MRtrix's `mrinfo` command to see whether a non-PyDesigner command withing NeuroDock is able to read this DICOM directory. If this doesn't work then I'll have to dive deeper into why the container is behaving this way. Run the following command and paste the output.

```
docker run -it --rm -v C:\Users\Tash\Onedrive\Documents:/data dmri/neurodock:latest mrinfo /data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20
```

**Reply by mmsarhan** [reply]
Im sorry for all the trouble. I am unsure why my computer is not working correctly. Here is the output when i used the command above.
   C:\Users\Tash\OneDrive\Documents>docker run -it --rm -v C:\Users\Tash\Onedrive\Documents:/data dmri/neurodock:latest mrinfo /data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20
mrinfo: [ERROR] unknown format for image "/data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20"
mrinfo: [ERROR] error opening image "/data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20"

C:\Users\Tash\OneDrive\Documents>

**Reply by TheJaeger** [reply]
That specific `mrinfo` error occurs when the path is invalid, meaning that `/data/DKI_BIPOLAR_2_5_mm_64dir_50slices_20` doesn't exist for the container. This could is most likely due to the mount not working as intended. I noticed that you're mounting `Onedrive` instead of `OneDrive`, perhaps the paths are case-sensitive and this could be throwing it off.

**Reply by TheJaeger** [reply]
Nevermind, I found the issue. There's an extra underscore in your path - it needs to be `2_5mm` instead of `2_5_mm`. Try running the PyDesigner command again:

```
docker run -it --rm -v C:\Users\Tash\Onedrive\Documents:/data dmri/neurodock:latest pydesigner --standard --output /data/output /data/DKI_BIPOLAR_2_5mm_64dir_50slices_20
```

**Comment by mmsarhan** [comment]
I think that fixed it the command ran. I am just pasting the output to make sure that it ran everything that it needed to. Does this output show that everything ran correctly?
C:\Users\Tash\OneDrive\Documents>docker run -it --rm -v C:\Users\Tash\Onedrive\Documents:/data dmri/neurodock:latest pydesigner --standard --output /data/output /data/DKI_BIPOLAR_2_5mm_64dir_50slices_20
mrinfo: [done] scanning DICOM folder "/data/DKI_...LAR_2_5mm_64dir_50slices_20"
mrinfo: [100%] reading DICOM series "DKI_BIPOLAR_2.5mm_64dir_50slices"
mrconvert -quiet /data/output/dwi0.mif /data/output/working.mif
Insufficient information in .json file to determine Fourier status. Assuming DWI is partial Fourier
[WARNING] Given DWI is partial fourier, overriding --degibbs; no unringing correction will be done to avoid artifacts.Use the "--adv" flag to run forced corrections.
dwifslpreproc: [WARNING] DWI header indicates no phase encoding contrast between b=0 images; proceeding without inhomogeneity field estimation
Plotting motion...
/usr/local/lib/python3.7/dist-packages/designer/plotting/motionplot.py:82: UserWarning: Tight layout not applied. The bottom and top margins cannot be made large enough to accommodate all axes decorations.
  plt.tight_layout()
mrmath: [100%] preloading data for "/data/output/B0_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/B0_MEAN.mif) ? /data/output/B0_MEAN.mif : 0)
mrmath: [100%] preloading data for "/data/output/b1000_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/b1000_MEAN.mif) ? /data/output/b1000_MEAN.mif : 0)
mrmath: [100%] preloading data for "/data/output/B0_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/B0_MEAN.mif) ? /data/output/B0_MEAN.mif : 0)
mrcalc: [WARNING] existing output files will be overwritten
mrcalc: [100%] computing: (/data/output/brain_mask.nii * /data/output/csf_mask.nii)
Running CSF-excluded smoothing at FWHM = 1.25...
Running Rician correction...
Path: /data/output
Name: dwirc
Full: /data/output/dwirc.nii
mrmath: [100%] preloading data for "/data/output/B0_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/B0_MEAN.mif) ? /data/output/B0_MEAN.mif : 0)
Computing SNR: dwi_raw.nii
   * B0...
   * B1000...
   * B2000...
Computing SNR: dwi_preprocessed.nii
   * B0...
   * B1000...
   * B2000...
Image dwi_preprocessed.nii loaded successfully
Processing with 4 workers...
Protocol(s) detected: DTI, DKI
IRLLS Noise Estimation: [100%]
IRLLS Outlier Detection: [100%]
/data/output/metrics_qc/fitting/outliers_irlls.nii
Plotting outliers...
Constrained Tensor Fit: [100%]
AKC Outlier Detection: [0%]
C:\Users\Tash\OneDrive\Documents>

**Reply by TheJaeger** [reply]
That looks like it ran fine.

**Reply by mmsarhan** [reply]
What about the last line "AKC Outlier Detection: [0%]" ? Is it normal that it did not run. Also, is there a guide to know how to look at the images from the data?

**Reply by TheJaeger** [reply]
No that's not normal. It should have gone to completion at 100%, followed by DTI, DKI and WMTI parameter extraction. I assumed you pasted a part of the output. This happens when you run out of memory.

AKC correction is highly demanding in terms of memory. This specific sequence occupies ~20 GB of RAM for AKC correction so I'd suggest allocating at least 16 GB of memory to your Docker containers. Processing can be further sped up by allocating more CPU cores to your containers. Docker's documentation covers how to do both of these on Windows systems.

I primarily use FSLeyes for visualization of output NifTi files. A VM running Ubuntu or headless Ubuntu installation on WSL will allow you to get it to work on Windows. Otherwise, tools like ImageJ Fiji or Mango also work.

**Reply by mmsarhan** [reply]
Sorry for taking so long to respond. We recently received a hard drive that I am trying to use for the processing to help with the space issue. Do I have to place the docker in the hard drive folder? Also, I have been trying to set up the path to mount the file in the hard drive but I keep receiving an error message. Is this the correct way to mount it on to the hard drive? The letter given to the hard drive is D
docker run -it --rm -v D:/DKE:/data dmri/neurodock:latest pydesigner --standard --output /data/output /data/DKI_BIPOLAR_2_5mm_64dir_50slices_20

**Reply by mmsarhan** [reply]
I was able to run the command using the hard drive as the mounted file but it did not go through the whole processing. I pasted the output below. Im not sure what went wrong.

C:\Users\Tash>docker run -it --rm -v D:\DKE:/data dmri/neurodock:latest pydesigner --standard --output /data/output /data/DKI_BIPOLAR_2_5mm_64dir_50slices_20
mrinfo: [done] scanning DICOM folder "/data/DKI_...LAR_2_5mm_64dir_50slices_20"
mrinfo: [100%] reading DICOM series "DKI_BIPOLAR_2.5mm_64dir_50slices"
mrconvert -quiet /data/output/dwi0.mif /data/output/working.mif
Insufficient information in .json file to determine Fourier status. Assuming DWI is partial Fourier
[WARNING] Given DWI is partial fourier, overriding --degibbs; no unringing correction will be done to avoid artifacts.Use the "--adv" flag to run forced corrections.
dwifslpreproc: [WARNING] DWI header indicates no phase encoding contrast between b=0 images; proceeding without inhomogeneity field estimation
Plotting motion...
/usr/local/lib/python3.7/dist-packages/designer/plotting/motionplot.py:82: UserWarning: Tight layout not applied. The bottom and top margins cannot be made large enough to accommodate all axes decorations.
  plt.tight_layout()
mrmath: [100%] preloading data for "/data/output/B0_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/B0_MEAN.mif) ? /data/output/B0_MEAN.mif : 0)
mrmath: [100%] preloading data for "/data/output/b1000_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/b1000_MEAN.mif) ? /data/output/b1000_MEAN.mif : 0)
mrmath: [100%] preloading data for "/data/output/B0_ALL.mif"
mrmath: [100%] computing mean along axis 3...
mrcalc: [100%] computing: (finite (/data/output/B0_MEAN.mif) ? /data/output/B0_MEAN.mif : 0)
mrcalc: [WARNING] existing output files will be overwritten
mrcalc: [ERROR] error creating output file "/data/output/csf_mask.nii": No such file or directory
mrcalc: [ERROR] error creating image "/data/output/csf_mask.nii"
Traceback (most recent call last):
  File "/usr/local/bin/pydesigner", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/designer/pydesigner.py", line 737, in main
    raise Exception('Unable to multiply CSF mask with brain '
Exception: Unable to multiply CSF mask with brain mask. See above for errors.

C:\Users\Tash>

**Reply by TheJaeger** [reply]
Hi @mmsarhan, you do not need to reinstall NeuroDock again since it resides on your host OS drive just like any other software.

I've never seen that error before but it appears that a certain preprocessing file was not created. Could you delete all contents of the output folder and rerun processing with the addition of a `--verbose` flag? Paste the output log here if it fails.

**Reply by mmsarhan** [reply]
I am not sure why, but when I run the command it does not work anymore. Even when I add the --verbose it does not work. I have both outputs with and without --verbose to show it does not work. Im not sure why its not working. 

C:\Users\Tash>docker run -it --rm -v D:\DKE:/data dmri/neurodock:latest pydesigner --standard --output /data/output /data/DKI_BIPOLAR_2_5mm_64dir_50slices_20
docker: Error response from daemon: error while creating mount source path '/run/desktop/mnt/host/d/DKE': mkdir /run/desktop/mnt/host/d: file exists.

C:\Users\Tash>docker run -it --rm -v D:\DKE:/data dmri/neurodock:latest pydesigner --standard --output --verbose /data/output /data/DKI_BIPOLAR_2_5mm_64dir_50slices_20
docker: Error response from daemon: error while creating mount source path '/run/desktop/mnt/host/d/DKE': mkdir /run/desktop/mnt/host/d: file exists.

C:\Users\Tash>
